espnet2/train/preprocessor.py

```diff
@@ -1340,3 +1340,179 @@ def __call__(
+    ) -> Tuple[Dict[str, Union[str, np.ndarray]], int]:
+        assert check_argument_types()
+
+        init_pad = 0
+        if self.speech_name in data:
+            speech = data[self.speech_name]
+
+            # speech: (Nmic, Time)
+            if speech.ndim == 1:
+                speech = speech[None, :]
+            else:
+                speech = speech.T
+
+            # Add silence to the left
+            if self.train and speech.shape[-1] < self.speech_length:
+                init_pad = np.random.randint(
+                    min(self.speech_length - speech.shape[-1], self.speech_init_silence)
+                    + 1
+                )
+                speech = np.pad(speech, ((0, 0), (init_pad, 0)))
+
+            # Pad or trim to max_samples
+            if speech.shape[-1] < self.speech_length:
+                speech = np.pad(
+                    speech, ((0, 0), (0, self.speech_length - speech.shape[-1]))
+                )
+            else:
+                speech = speech[:, : self.speech_length]
+
+            data[self.speech_name] = speech.T  # convert back to time first
+
+        assert check_return_type((data, init_pad))
+        return data, init_pad
+
+    def _text_process(
+        self, data: Dict[str, Union[str, np.ndarray]], time_shift: int
+    ) -> Dict[str, np.ndarray]:
+        assert check_argument_types()
+
+        text_names = [self.text_name, self.text_prev_name, self.text_ctc_name]
+        if self.tokenizer is not None:
+            for name in text_names:
+                if name in data:
+                    text = data[name]
+
+                    # Remove prev text by setting it to <na>
+                    if (
+                        self.train
+                        and name == self.text_prev_name
+                        and np.random.uniform() > self.text_prev_apply_prob
+                    ):
+                        text = self.na_symbol
+
+                    text = self.text_cleaner(text)
+                    tokens = self.tokenizer.text2tokens(text)
+                    text_ints = self.token_id_converter.tokens2ids(tokens)
+                    text_ints = np.array(text_ints, dtype=np.int64)
+
+                    # Augment text
+                    if name == self.text_name:
+                        # NOTE(yifan): The first token is always space which should be removed,
+                        # because no space should be present between special tokens.
+                        # This works for bpe, but probably not for the other token types.
+                        text_ints = text_ints[1:]
+
+                        # Remove timestamps
+                        if self.train and np.random.uniform() > self.time_apply_prob:
+                            # Timestamps are continuous ints
+                            text_ints = text_ints[
+                                np.logical_or(
+                                    text_ints < self.first_time,
```

Comment:
This may not work for `char` token types. Two main problems I noticed: 1. the order of tokens is following the frequency, not strictly continuous. 2. if a token does not occur in the text, it will be skipped in the token list. To address the 1st problem, maybe it is safer to use
```
self.all_time_ints = np.array(
      [self.token_id_converter.token2id[x] for x in all_time_tokens]
)
time_indices = np.in1d(text_ints, self.all_time_ints)
text_ints = text_ints[~time_indices]
```

Reply:
Thanks for the comment! Currently, I think only the `bpe` token type can work from training to inference. I can add a type check in the `s2t.sh` which terminates the program if any other token is specified.

Also, speed perturbation is not supported because the timestamps will change.